templates/node/node.py

```diff
@@ -0,0 +1,269 @@
+            'port_forwards':portforwards, 
+            'protocol':'tcp'})
+        task.wait()
+
+    def portforward_delete(self, portforwards={}):
+        """ Delete portforwards """
+
+        machine = self.machine
+        machine_ip, _ = machine.machineip_get()
+
+        space = self.space
+        vdc = self.vdc
+        task = vdc.schedule_action(
+            'portforward_delete', 
+            {'machineId':machine.id, 
+            'port_forwards':self.data['ports'], 
+            'protocol':'tcp'})
+        task.wait()        
+ 
+
+    def start(self):
+        """ Start the VM """
+
+        machine = self.machine
+        if machine:
+            machine.start()
+        else:
+            raise RuntimeError('machine %s in not found' % self.name)
+
+    def stop(self):
+        """ Stop the VM """
+
+        machine = self.machine
+        if machine:
+            machine.stop()
+        else:
+            raise RuntimeError('machine %s in not found' % self.name)
+
+    def restart(self):
+        """ Restart the VM """
+
+        machine = self.machine
+        if machine:
+            machine.restart()
+        else:
+            raise RuntimeError('machine %s in not found' % self.name)
+
+    def pause(self):
+        """ Pause the VM """
+
+        machine = self.machine
+        if machine:
+            machine.pause()
+        else:
+            raise RuntimeError('machine %s in not found' % self.name)
+
+    def resume(self):
+        """ Resume the VM """
+
+        machine = self.machine
+        if machine:
+            machine.resume()
+        else:
+            raise RuntimeError('machine %s in not found' % self.name)
+
+    def reset(self):
+        """ Reset the VM """
+
+        machine = self.machine
+        if machine:
+            machine.reset()    
+        else:
+            raise RuntimeError('machine %s in not found' % self.name)      
+
+    def snapshot(self):
+        """
+        Action that creates a snapshot of the machine
+        """
+        machine = self.machine
+        if machine:
+            machine.snapshot_create()
+        else:
+            raise RuntimeError('machine %s in not found' % self.name)
+
+    def snapshot_rollback(self, snapshot_epoch=None):
+        """
+        Action that rolls back the machine to a snapshot
+        """    
+        if not snapshot_epoch:
+            raise RuntimeError('"snapshot_epoch" should be given')
+
+        machine = self.machine
+        if machine:       
+            machine.snapshot_rollback(snapshot_epoch)
+            machine.start()
+        else:
+            raise RuntimeError('machine %s in not found' % self.name)
+
+    def snapshot_delete(self,  snapshot_epoch=None):
+        """
+        Action that deletes a snapshot of the machine
+        """
+        if not snapshot_epoch:
+            raise RuntimeError('"snapshot_epoch" should be given')
+
+        machine = self.machine
+        if machine and snapshot_epoch:       
```

Comment:
test machine and snapshot epoch in 2 conditions, so you can raise with proper error message for both